2-onchain-interactions.md

```diff
@@ -0,0 +1,284 @@
+
+QUESTIONS:
+
+- What happens if A (not turn taker) wants to withdraw funds (B is malicious). They would put state n-1 onchain with `forceMove`, and B would be free to call `respond` to clear the challenge. Is the `ChallengeCleared` event verbose enough for A to create a new state somehow? Is A expected to get the data B used from the calldata?
+
+### Concluded
+
+The `Concluded` event is defined as:
+
+```typescript
+event Concluded(
+  // channelId with finalized outcomes
+  bytes32 indexed channelId,
+);
+```
+
+This event is emitted at the end of the `conclude` as well as the `concludePushOutcomeAndTransferAll` methods. Once the challenge has expired, the outcomes must be finalized before funds can be withdrawn, `concludePushOutcomeAndTransferAll` is a helper that executes the entire process in one transaction.
+
+## Design Requirements
+
+The onchain service must:
+
+- monitor the chain for channel events
+- push event information to the channel wallet for relevant channels
+- store information related to relevant channel events
+- send transactions with sent to the service from the channel wallet
+
+### Implementation Details
+
+To build iteratively, the implementation should be broken out into three phases:
+
+- **Phase 0**: stateless chain watching and transaction sending
+- **Phase 1**: stateful chain watching
+- **Phase 2**: lightweight, stateful chain watching with historical chain syncing (could include integration of outside tools, like a [subgraph](https://thegraph.com/docs/introduction#how-the-graph-works) or [dagger](https://github.com/maticnetwork/dagger.js))
+
+## Interface
+
+```typescript
+import { providers, Wallet } from "ethers";
+
+type OnchainServiceEnvironment = {
+  provider: string | providers.JsonRpcProvider;
+  wallet: Wallet;
+  acceptableErrors?: string[]; // Allow channel wallet to define errors for retries
+  maximumRetries?: number; // Maximum number of times a tx is retried
+};
+
+type MinimumTransaction = {
+  to: Address;
+  value: Uint256;
+  data: string; // calldata
+};
+
+type TransactionResult = providers.TransactionResponse & {
+  // TODO: this type of callback is useful if you need to wait not
+  // only for the transaction but also for some other transaction
+  // related asynchronous behavior (i.e. writing event data to storage)
+  completed: () => providers.TransactionReceipt & ContractEvent;
+};
+
+// TODO: why not make these the same as the contract event names where
+// applicable?
+type ChainEventName =
+  | "fundsDeposited"
+  | "fundsWithdrawn"
+  | "transactionMined"
+  | "forceMoveDetected"
+  | "checkpointDetected"
+  | "concludeDetected"
+  | "respondDetected";
```

Comment:
Seems weird to need to redefine these. You can use the `@statechannels/nitro-protocol` `npm` package I think